agent.go

```diff
@@ -0,0 +1,157 @@
+package wflambda
+
+import (
+	"log"
+	"os"
+
+	wavefront "github.com/wavefronthq/wavefront-sdk-go/senders"
+)
+
+var (
+	// Is this a cold start or not.
+	coldStart = true
+	// Count the number of cold starts.
+	csCounter = counter{}
+	// Count the number of invocations.
+	invocationsCounter = counter{}
+	// Count the number of errors.
+	errCounter = counter{}
+)
+
+// WavefrontConfig configures the direct ingestion sender to Wavefront.
+type WavefrontConfig struct {
+	// Enabled indicates whether metrics are sent to Wavefront
+	Enabled *bool
+	// Wavefront URL of the form https://<INSTANCE>.wavefront.com.
+	Server *string
+	// Wavefront API token with direct data ingestion permission.
+	Token *string
+	// Max batch of data sent per flush interval.
+	BatchSize *int
+	// Max size of internal buffers beyond which received data is dropped.
+	MaxBufferSize *int
+	// Map of Key-Value pairs (strings) associated with each data point sent to Wavefront.
+	PointTags map[string]string
+}
+
+// WavefrontAgent is the agent instance that communicates with Wavefront.
+type WavefrontAgent struct {
+	*WavefrontConfig
+	metrics  map[string]float64
+	counters map[string]float64
+	sender   wavefront.Sender
+}
+
+var (
+	// Default value whether the agent is enabled or not.
+	defaultEnabled = true
+	// Default value for batch of data sent per flush interval.
+	defaultBatchSize = 10000
+	// Default size of internal buffers beyond which received data is dropped
+	defaultMaxBufferSize = 50000
+	// Default interval (in seconds) at which to flush data to Wavefront.
+	defaultFlushIntervalSeconds = 1
+)
+
+// NewWavefrontAgent returns a new agent.
+func NewWavefrontAgent(w *WavefrontConfig) *WavefrontAgent {
+	// Create a new instance of the WavefrontAgent.
+	wfAgent := &WavefrontAgent{
+		metrics:  make(map[string]float64),
+		counters: make(map[string]float64),
+	}
+
+	// Create an empty map of point tags if no tags exist yet.
+	if w.PointTags == nil {
+		w.PointTags = make(map[string]string)
+	}
+
+	// Create the configuration to connect to Wavefront. Details are gathered from both
+	// the WavefrontConfig and the environment variables. If both WavefrontConfig and
+	// environment variables have a value for a specific setting, the environment variable
+	// takes precedence.
+	enabled := &defaultEnabled
+	envEnabled := os.Getenv("WAVEFRONT_ENABLED")
+	if w.Enabled != nil {
+		enabled = w.Enabled
+	}
+	if envEnabled != "" {
+		enabled = stringToBool(envEnabled)
+	}
+
+	envServer := os.Getenv("WAVEFRONT_URL")
```

Comment:
if "Enabled" , but "wavefront_url or wavefront_api_token" is unset, we can fail early.